x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizer.java

```diff
@@ -89,62 +80,26 @@ static class ProjectAwayColumns extends Rule<PhysicalPlan, PhysicalPlan> {
 
         @Override
         public PhysicalPlan apply(PhysicalPlan plan) {
-            var projectAll = new Holder<>(TRUE);
-            var keepCollecting = new Holder<>(TRUE);
-            var attributes = new AttributeSet();
-            var aliases = new AttributeMap<Expression>();
+            Holder<Boolean> keepTraversing = new Holder<>(TRUE);
+            // Invariant: if we add a projection with these attributes after the current plan node, the plan remains valid
+            // and the overall output will not change.
+            Holder<AttributeSet> requiredAttributes = new Holder<>(plan.outputSet());
```

Comment:
I need the holder because I need to re-assign the `requiredAttributes` after doing some non-mutating set operations like `subtract` and `combine`. Getting rid of the Holder would work if I only needed to mutate.